dataset-mover/src/main/scala/com/socrata/datacoordinator/mover/Main.scala

```diff
@@ -0,0 +1,490 @@
+    System.err.println()
+    System.err.println("  INTERNAL_NAME   internal name (e.g., alpha.1234) of the dataset to move")
+    System.err.println("  TARGET_TRUTH    truthstore in which to move it (e.g., bravo)")
+    System.err.println()
+    System.err.println("Unless the SOCRATA_COMMIT_MOVE environment variable is set, all changes")
+    System.err.println("will be rolled back rather than committed.")
+    sys.exit(1)
+  }
+
+  val fromInternalName = DatasetInternalName(args(0)).getOrElse {
+    System.err.println("Illegal dataset internal name")
+    sys.exit(1)
+  }
+  val fromDatasetId = fromInternalName.datasetId
+
+  val toInstance = args(1)
+
+  val dryRun = sys.env.get("SOCRATA_COMMIT_MOVE").isEmpty
+
+  val serviceConfig = try {
+    new MoverConfig(ConfigFactory.load(), "com.socrata.coordinator.datasetmover")
+  } catch {
+    case e: Exception =>
+      Console.err.println(e)
+      sys.exit(1)
+  }
+
+  PropertyConfigurator.configure(Propertizer("log4j", serviceConfig.logProperties))
+
+  val log = org.slf4j.LoggerFactory.getLogger(classOf[Main])
+
+  implicit val executorShutdown = Resource.executorShutdownNoTimeout
+
+  def fullyReplicated(datasetId: DatasetId, manifest: SecondaryManifest, targetVersion: Long): Boolean = {
+    manifest.stores(datasetId).values.forall(_ == targetVersion)
+  }
+
+  def isPgSecondary(store: String): Boolean =
+    serviceConfig.pgSecondaries.contains(store)
+
+  if(serviceConfig.truths.values.exists(_.poolOptions.isDefined)) {
+    System.err.println("truths must not be c3p0 data sources")
+    sys.exit(1)
+  }
+
+  def doCopy(fromConn: Connection, copyOutSql: String, toConn: Connection, copyInSql: String) {
+    log.info("Copy from: {}", copyOutSql)
+    log.info("Copy to: {}", copyInSql)
+    val fromCopyOut = fromConn.asInstanceOf[PGConnection].getCopyAPI.copyOut(copyOutSql)
+    try {
+      val toCopyIn = toConn.asInstanceOf[PGConnection].getCopyAPI.copyIn(copyInSql)
+      try {
+        // This buffering is WAY faster than just writing the chunks we get straight back out
+        val buffer = new Array[Byte](1024*1024)
+        var bufferEnd = 0
+
+        @tailrec
+        def loop(total: Long, lastWrote: Long) {
+          val chunk = fromCopyOut.readFromCopy()
+          if(chunk ne null) {
+            var written = 0L
+            if(chunk.length + bufferEnd > buffer.length && bufferEnd != 0) {
+              toCopyIn.writeToCopy(buffer, 0, bufferEnd)
+              written += bufferEnd
+              bufferEnd = 0
+            }
+            if(chunk.length > buffer.length) {
+              toCopyIn.writeToCopy(chunk, 0, chunk.length)
+              written += chunk.length
+            } else {
+              System.arraycopy(chunk, 0, buffer, bufferEnd, chunk.length)
+              bufferEnd += chunk.length
+            }
+            val newTotal = total + written
+            val now = System.nanoTime()
+            if(now - lastWrote > 1000000000L) {
```

Comment:
I think this is the weirdest way to express "1 second has passed" I've ever seen, but I guess it works.

Reply:
Fun fact: System.nanoTime() has no epoch, so only differences between two values acquired within the same process has a meaning (in particular: it's not guaranteed that a later timestamp will be "greater than" an earlier timestamp, only that the difference between them will be the difference in nanoseconds up to wraparound)